File: YTCnv/Platforms/Android/DownloadNotificationService.cs
using Android.App;
using Android.Content;
using Android.Content.PM;
using Android.OS;
using AndroidX.Core.App;

namespace YTCnv
{
    [Service(Name = "com.pg_axis.downloadnotificationservice", Exported = true, ForegroundServiceType = ForegroundService.TypeDataSync)]
    public class DownloadNotificationService : Service
    {
        private const int NOTIFICATION_ID = 1001;
        private const string CHANNEL_ID = "download_channel";

        public override void OnCreate()
        {
            base.OnCreate();
            CreateNotificationChannel();
        }

        public override void OnDestroy()
        {
            base.OnDestroy();
            StopForeground(true);
        }

        public override StartCommandResult OnStartCommand(Intent intent, StartCommandFlags flags, int startId)
        {
            var notification = BuildNotification();
            StartForeground(NOTIFICATION_ID, notification);
            return StartCommandResult.Sticky;
        }

        public override IBinder OnBind(Intent intent)
        {
            return null;
        }

        private Notification BuildNotification()
        {
            var intent = new Intent(this, typeof(MainActivity));
            intent.SetFlags(ActivityFlags.ClearTop | ActivityFlags.SingleTop);

            var pendingIntent = PendingIntent.GetActivity(
                this,
                0,
                intent,
                PendingIntentFlags.Immutable
            );

            var builder = new NotificationCompat.Builder(this, CHANNEL_ID)
                .SetContentTitle("Downloading in background")
                .SetContentText("The download is running. Please keep the app open.")
                .SetSmallIcon(Resource.Drawable.icon)
                .SetColor(Android.Graphics.Color.ParseColor("#2b950f"))
                .SetContentIntent(pendingIntent)
                .SetOngoing(true)
                .SetPriority(NotificationCompat.PriorityLow)
                .SetVisibility(NotificationCompat.VisibilityPublic);

            return builder.Build();
        }

        private void CreateNotificationChannel()
        {
            if (Build.VERSION.SdkInt >= BuildVersionCodes.O)
            {
                var channel = new NotificationChannel(
                    CHANNEL_ID,
                    "Download Background Service",
                    NotificationImportance.Low
                )
                {
                    Description = "Handles background downloading service"
                };

                var manager = (NotificationManager)GetSystemService(NotificationService);
                manager.CreateNotificationChannel(channel);
            }
        }
    }

}

File: app/src/main/java/com/pg_axis/ytcnv/DownloadNotificationService.kt
package com.pg_axis.ytcnv

import android.app.Notification
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.app.Service
import android.content.Context
import android.content.Intent
import android.os.IBinder

class DownloadNotificationService : Service() {

    companion object {
        const val CHANNEL_ID = "ytcnv_download_channel"
        const val FINISH_CHANNEL_ID = "ytcnv_finnish_channel"
        const val NOTIFICATION_ID = 1
        const val FINISH_NOTIFICATION_ID = 2

        fun showFinishNotification(context: Context, fileName: String) {
            val manager = context.getSystemService(NotificationManager::class.java)

            val intent = Intent(context, MainActivity::class.java).apply {
                flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK
            }
            val pendingIntent = PendingIntent.getActivity(
                context, 0, intent, PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE
            )

            val notification = Notification.Builder(context, FINISH_CHANNEL_ID)
                .setContentTitle("Download Finished")
                .setContentText("Downloaded $fileName")
                .setSmallIcon(R.drawable.splash)
                .setContentIntent(pendingIntent)
                .setAutoCancel(true)
                .build()

            manager.notify(FINISH_NOTIFICATION_ID, notification)
        }
    }

    override fun onCreate() {
        super.onCreate()

        val manager = getSystemService(NotificationManager::class.java)

        manager.createNotificationChannel(
            NotificationChannel(
                CHANNEL_ID,
                "YTCnv Downloads",
                NotificationManager.IMPORTANCE_LOW
            ).apply { description = "Download progress notifications" }
        )

        manager.createNotificationChannel(
            NotificationChannel(
                FINISH_CHANNEL_ID,
                "YTCnv Download Complete",
                NotificationManager.IMPORTANCE_DEFAULT
            ).apply { description = "Notifies when a download finishes" }
        )
    }


    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        val openIntent = Intent(this, MainActivity::class.java)
        openIntent.flags = Intent.FLAG_ACTIVITY_SINGLE_TOP or Intent.FLAG_ACTIVITY_REORDER_TO_FRONT

        val pendingIntent = PendingIntent.getActivity(
            this, 0, openIntent,
            PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE
        )

        val notification = Notification.Builder(this, CHANNEL_ID)
            .setContentTitle("YTCnv")
            .setContentText("Download in progress...")
            .setSmallIcon(R.drawable.splash)
            .setOngoing(true)
            .setContentIntent(pendingIntent)
            .build()

        startForeground(NOTIFICATION_ID, notification)
        return START_NOT_STICKY
    }

    override fun onBind(intent: Intent?): IBinder? = null
}